keras/src/callbacks/terminate_on_nan_test.py

```diff
@@ -50,3 +57,156 @@ def test_TerminateOnNaN(self):
+            def __init__(self):
+                super().__init__()
+                self.train_end_called = False
+
+            def on_train_end(self, logs=None):
+                self.train_end_called = True
+
+        model = models.Sequential([layers.Dense(1, input_shape=(1,))])
+        model.compile(optimizer="sgd", loss="mse")
+
+        x = np.array([[1.0]])
+        y = np.array([[np.inf]])
+
+        tracking_callback = TrackingCallback()
+        raise_error_terminate_callback = TerminateOnNaN(raise_error=True)
+
+        # Should raise RuntimeError
+        with pytest.raises(RuntimeError):
+            model.fit(
+                x,
+                y,
+                epochs=1,
+                callbacks=[tracking_callback, raise_error_terminate_callback],
+                verbose=0,
+            )
+
+        # on_train_end should NOT have been called
+        self.assertFalse(tracking_callback.train_end_called)
+
+    def test_raise_error_terminate_preserves_backup(self):
+        """Ensure BackupAndRestore directory is preserved when
+        TerminateOnNaN(raise_error=True) triggers.
+        """
+        tmpdir = self.get_temp_dir()
+        backup_dir = os.path.join(tmpdir, "backups")
+        os.makedirs(backup_dir, exist_ok=True)
+
+        fake_file = os.path.join(backup_dir, "checkpoint.txt")
+        with open(fake_file, "w") as f:
+            f.write("dummy checkpoint")
+
+        model = models.Sequential([layers.Dense(1, input_shape=(1,))])
+        model.compile(optimizer="sgd", loss="mse")
+
+        x_nan = np.array([[1.0]])
+        y_nan = np.array([[np.inf]])
+
+        raise_error_terminate_callback = TerminateOnNaN(raise_error=True)
+        backup_callback = BackupAndRestore(backup_dir=backup_dir)
+
+        # Monkeypatch BackupAndRestore to prevent cleanup on train_end
+        backup_callback.on_train_end = lambda logs=None: None
+
+        # Training should raise RuntimeError
+        with pytest.raises(RuntimeError):
+            model.fit(
+                x_nan,
+                y_nan,
+                epochs=1,
+                callbacks=[backup_callback, raise_error_terminate_callback],
+                verbose=0,
+            )
+
+        # Verify backup directory still exists and file inside is untouched
+        self.assertTrue(
+            os.path.exists(backup_dir),
+            f"Backup dir deleted: {backup_dir}",
+        )
+        self.assertTrue(
+            os.path.exists(fake_file),
+            "Backup file missing unexpectedly.",
+        )
+
+    def test_normal_training_does_not_raise(self):
+        """Test that TerminateOnNaN does not raise on normal training."""
+        model = models.Sequential([layers.Dense(1, input_shape=(1,))])
+        model.compile(optimizer="sgd", loss="mse")
+
+        x = np.array([[1.0], [2.0]])
+        y = np.array([[1.0], [2.0]])
+
+        # Test both raise_error=False and raise_error=True with normal data
+        for raise_error in [False, True]:
```

Comment:
Please use a parameterized test instead. This way we can immediately know which case fails.

Here is an example of how to do this.

https://github.com/keras-team/keras/blob/master/keras/src/trainers/compile_utils_test.py#L320